File: src/main/java/com/kiroule/javacodingskills/codility/lesson2/CyclicRotation.java
package com.kiroule.javacodingskills.codility.lesson2;

/**
 * @author Igor Baiborodine
 *
 * @see ../CYCLIC_ROTATION_README.md
 */
public class CyclicRotation {

  public int[] solution(int[] A, int K) {
    if (A.length <= 1
        || (K >= A.length && K % A.length == 0)) {
      return A;
    }
    if (K > A.length) {
      K = K % A.length;
    }

    rotations:
    for (int i = 0; i < K; i++) {
      int lastElement = A[A.length - 1];
      elements:
      for (int j = A.length - 1; j > 0; j--) {
        A[j] = A[j - 1];
      }
      A[0] = lastElement;
    }
    return A;
  }

}

File: codility/lesson2/CyclicRotation.py
"""Solution for the Cyclic Rotation exercise.

Exercise: CYCLIC_ROTATION_README.md
Author:   Igor Baiborodine

"""


def solution(a, k):
    if len(a) <= 1 or (k >= len(a) and k % len(a) == 0):
        return a
  
    if k > len(a):
        k = k % len(a)
  
    while k > 0:
        last_elem = a[len(a) - 1]
        a.insert(0, last_elem)
        del a[len(a) - 1]
        k -= 1

    return a


assert solution([], 0) == []
assert solution([1], 0) == [1]
assert solution([1, 2, 3, 4, 5], 5) == [1, 2, 3, 4, 5]
assert solution([1, 2, 3, 4, 5], 10) == [1, 2, 3, 4, 5]
assert solution([1, 2, 3, 4], 2) == [3, 4, 1, 2]
assert solution([1, 2, 3, 4], 3) == [2, 3, 4, 1]
assert solution([1, 2, 3, 4], 6) == [3, 4, 1, 2]
assert solution([3, 8, 9, 7, 6], 3) == [9, 7, 6, 3, 8]
